File: scripts/recover_from_transcript.py
#!/usr/bin/env python3
"""Recover files from Cursor agent transcript JSONL by replaying Write/StrReplace ops."""

from __future__ import annotations

import json
import os
import re
import sys
from pathlib import Path

TRANSCRIPT = Path(
    r"C:\Users\laksh\.cursor\projects\d-Work-Projects-Latheif-Productions-Pixel-Pluz-Software-pixlpluzportal"
    r"\agent-transcripts\ca00464f-795f-4a27-bf61-556dff566421\ca00464f-795f-4a27-bf61-556dff566421.jsonl"
)
WORKSPACE = Path(r"d:\Work\Projects\Latheif Productions\Pixel Pluz\Software\pixlpluzportal")

EXPLICIT_TARGETS = {
    "app/api/admin/tasks/route.ts",
    "app/api/admin/task-submissions/route.ts",
    "app/api/admin/class-materials/route.ts",
    "app/api/admin/attendance/route.ts",
    "app/api/admin/complaints/route.ts",
    "app/api/admin/mentor-ratings/route.ts",
    "app/task-submissions/submit/[taskId]/page.tsx",
}

STUDENT_PREFIX = "app/api/student/"

NO_RESUBMIT_PATHS = {
    "app/api/admin/task-submissions/route.ts",
    "app/task-submissions/submit/[taskId]/page.tsx",
}

RESUBMIT_MARKERS = (
    "handleResubmit",
    "canStudentResubmit",
    "resubmitTask",
    "submission_history",
    "resubmit_deadline",
    "studentCanResubmit",
    "SubmissionHistoryEntry",
    "useSearchParams",
    "type: 'initial' | 'resubmit'",
    "isResubmitMode",
    "resubmit mode",
)


def normalize_path(raw: str) -> str | None:
    if not raw:
        return None
    p = raw.replace("\\", "/")
    prefix = "d:/work/projects/latheif productions/pixel pluz/software/pixlpluzportal/"
    if p.lower().startswith(prefix):
        p = p[len(prefix) :]
    m = re.search(r"pixlpluzportal[/\\](.+)$", p, re.I)
    if m:
        p = m.group(1)
    return p.replace("\\", "/").lstrip("/")


def is_target_path(rel: str) -> bool:
    if rel in EXPLICIT_TARGETS:
        return True
    return rel.startswith(STUDENT_PREFIX) and rel.endswith(".ts")


def has_resubmit_content(content: str) -> bool:
    return any(marker in content for marker in RESUBMIT_MARKERS)


def apply_str_replace(content: str | None, old: str, new: str) -> str | None:
    if content is None:
        return None
    if old not in content:
        return content
    return content.replace(old, new, 1)


def is_revert_shell_line(obj: dict) -> bool:
    try:
        for block in obj.get("message", {}).get("content", []):
            if block.get("type") != "tool_use" or block.get("name") != "Shell":
                continue
            cmd = block.get("input", {}).get("command", "")
            if "Remove-Item" in cmd and "app/api/admin/tasks" in cmd:
                return True
    except Exception:
        pass
    return False


def extract_tool_ops(obj: dict, line_no: int) -> list[dict]:
    ops: list[dict] = []
    content = obj.get("message", {}).get("content", [])
    if not isinstance(content, list):
        return ops
    for block in content:
        if block.get("type") != "tool_use":
            continue
        name = block.get("name")
        if name not in ("Write", "StrReplace"):
            continue
        inp = block.get("input") or {}
        path = normalize_path(inp.get("path", ""))
        if not path or not is_target_path(path):
            continue
        ops.append(
            {
                "line": line_no,
                "tool": name,
                "path": path,
                "contents": inp.get("contents"),
                "old_string": inp.get("old_string"),
                "new_string": inp.get("new_string"),
            }
        )
    return ops


def main() -> int:
    if not TRANSCRIPT.exists():
        print(f"ERROR: Transcript not found: {TRANSCRIPT}", file=sys.stderr)
        return 1

    files: dict[str, str | None] = {}
    no_resubmit: dict[str, str] = {}
    last_line: dict[str, int] = {}
    all_ops: list[dict] = []
    revert_line: int | None = None

    with TRANSCRIPT.open("r", encoding="utf-8") as f:
        for line_no, line in enumerate(f, start=1):
            line = line.strip()
            if not line:
                continue
            try:
                obj = json.loads(line)
            except json.JSONDecodeError:
                continue

            if is_revert_shell_line(obj):
                revert_line = line_no
                break

            all_ops.extend(extract_tool_ops(obj, line_no))

    for op in all_ops:
        path = op["path"]
        if op["tool"] == "Write" and op["contents"] is not None:
            files[path] = op["contents"]
            last_line[path] = op["line"]
            if path in NO_RESUBMIT_PATHS and not has_resubmit_content(op["contents"]):
                no_resubmit[path] = op["contents"]
        elif op["tool"] == "StrReplace":
            old = op.get("old_string")
            new = op.get("new_string")
            if old is not None and new is not None:
                updated = apply_str_replace(files.get(path), old, new)
                if updated is not None:
                    files[path] = updated
                    last_line[path] = op["line"]
                    if path in NO_RESUBMIT_PATHS and not has_resubmit_content(updated):
                        no_resubmit[path] = updated

    discovered = sorted({op["path"] for op in all_ops if op["path"].startswith(STUDENT_PREFIX)})
    targets = set(EXPLICIT_TARGETS) | set(discovered)

    written: list[str] = []
    failed: list[str] = []
    skipped_resubmit: list[str] = []

    for rel in sorted(targets):
        if rel in NO_RESUBMIT_PATHS and rel in no_resubmit:
            content = no_resubmit[rel]
            if rel in files and files[rel] and has_resubmit_content(files[rel]):
                skipped_resubmit.append(rel)
        elif rel in files and files[rel] is not None:
            content = files[rel]
        else:
            failed.append(rel)
            continue

        out_path = WORKSPACE / rel.replace("/", os.sep)
        out_path.parent.mkdir(parents=True, exist_ok=True)
        out_path.write_text(content.replace("\r\n", "\n"), encoding="utf-8", newline="\n")
        written.append(rel)

    print("=== Recovery Report ===")
    print(f"Transcript lines processed: {revert_line or 'all'}")
    print(f"Total tool ops matched: {len(all_ops)}")
    print(f"Discovered student routes: {discovered or '(none)'}")
    print()
    print(f"Successfully written ({len(written)}):")
    for p in written:
        note = " [used pre-resubmit version]" if p in skipped_resubmit else ""
        print(f"  - {p} (last op ~line {last_line.get(p, '?')}){note}")
    print()
    if failed:
        print(f"Could NOT recover ({len(failed)}):")
        for p in failed:
            print(f"  - {p}")
    else:
        print("Could NOT recover: (none)")

    return 0 if not failed else 2


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/recover_from_transcript.mjs
#!/usr/bin/env node
/** Recover files from Cursor agent transcript JSONL by replaying Write/StrReplace ops. */

import fs from 'fs'
import path from 'path'
import readline from 'readline'

const TRANSCRIPT = path.join(
  'C:',
  'Users',
  'laksh',
  '.cursor',
  'projects',
  'd-Work-Projects-Latheif-Productions-Pixel-Pluz-Software-pixlpluzportal',
  'agent-transcripts',
  'ca00464f-795f-4a27-bf61-556dff566421',
  'ca00464f-795f-4a27-bf61-556dff566421.jsonl',
)
const WORKSPACE = path.join(
  'd:',
  'Work',
  'Projects',
  'Latheif Productions',
  'Pixel Pluz',
  'Software',
  'pixlpluzportal',
)

const EXPLICIT_TARGETS = new Set([
  'app/api/admin/tasks/route.ts',
  'app/api/admin/task-submissions/route.ts',
  'app/api/admin/class-materials/route.ts',
  'app/api/admin/attendance/route.ts',
  'app/api/admin/complaints/route.ts',
  'app/api/admin/mentor-ratings/route.ts',
  'app/task-submissions/submit/[taskId]/page.tsx',
])

const STUDENT_PREFIX = 'app/api/student/'

const NO_RESUBMIT_PATHS = new Set([
  'app/api/admin/task-submissions/route.ts',
  'app/task-submissions/submit/[taskId]/page.tsx',
])

const RESUBMIT_MARKERS = [
  'handleResubmit',
  'canStudentResubmit',
  'resubmitTask',
  'submission_history',
  'resubmit_deadline',
  'studentCanResubmit',
  'SubmissionHistoryEntry',
  'useSearchParams',
  "type: 'initial' | 'resubmit'",
  'isResubmitMode',
  'resubmit mode',
]

function normalizePath(raw) {
  if (!raw) return null
  let p = raw.replace(/\\/g, '/')
  const prefixes = [
    'd:/work/projects/latheif productions/pixel pluz/software/pixlpluzportal/',
  ]
  const lower = p.toLowerCase()
  for (const prefix of prefixes) {
    if (lower.startsWith(prefix)) {
      p = p.slice(prefix.length)
      break
    }
  }
  const m = p.match(/pixlpluzportal[/\\](.+)$/i)
  if (m) p = m[1]
  return p.replace(/\\/g, '/').replace(/^\//, '')
}

function isTargetPath(rel) {
  if (EXPLICIT_TARGETS.has(rel)) return true
  if (rel.startsWith(STUDENT_PREFIX) && rel.endsWith('.ts')) return true
  return false
}

function hasResubmitContent(content) {
  return RESUBMIT_MARKERS.some((m) => content.includes(m))
}

function applyStrReplace(content, oldStr, newStr) {
  if (content == null) return null
  if (!content.includes(oldStr)) return content
  return content.replace(oldStr, newStr)
}

function isRevertShellLine(obj) {
  try {
    const content = obj?.message?.content
    if (!Array.isArray(content)) return false
    for (const block of content) {
      if (block?.type !== 'tool_use' || block?.name !== 'Shell') continue
      const cmd = block?.input?.command || ''
      if (cmd.includes('Remove-Item') && cmd.includes('app/api/admin/tasks')) return true
    }
  } catch {
    /* ignore */
  }
  return false
}

function extractToolOps(obj, lineNo) {
  const ops = []
  const content = obj?.message?.content
  if (!Array.isArray(content)) return ops
  for (const block of content) {
    if (block?.type !== 'tool_use') continue
    const name = block?.name
    if (name !== 'Write' && name !== 'StrReplace') continue
    const inp = block?.input || {}
    const rel = normalizePath(inp.path || '')
    if (!rel || !isTargetPath(rel)) continue
    ops.push({
      line: lineNo,
      tool: name,
      path: rel,
      contents: inp.contents,
      old_string: inp.old_string,
      new_string: inp.new_string,
    })
  }
  return ops
}

async function main() {
  if (!fs.existsSync(TRANSCRIPT)) {
    console.error(`ERROR: Transcript not found: ${TRANSCRIPT}`)
    process.exit(1)
  }

  /** @type {Record<string, string|null>} */
  const files = {}
  /** @type {Record<string, string>} */
  const noResubmit = {}
  /** @type {Record<string, number>} */
  const lastLine = {}
  /** @type {Array<object>} */
  const allOps = []
  let revertLine = null

  const rl = readline.createInterface({
    input: fs.createReadStream(TRANSCRIPT, { encoding: 'utf8' }),
    crlfDelay: Infinity,
  })

  let lineNo = 0
  for await (const line of rl) {
    lineNo++
    const trimmed = line.trim()
    if (!trimmed) continue
    let obj
    try {
      obj = JSON.parse(trimmed)
    } catch {
      continue
    }

    if (isRevertShellLine(obj)) {
      revertLine = lineNo
      break
    }

    for (const op of extractToolOps(obj, lineNo)) {
      allOps.push(op)
    }
  }

  for (const op of allOps) {
    const rel = op.path
    if (op.tool === 'Write' && op.contents != null) {
      files[rel] = op.contents
      lastLine[rel] = op.line
      if (NO_RESUBMIT_PATHS.has(rel) && !hasResubmitContent(op.contents)) {
        noResubmit[rel] = op.contents
      }
    } else if (op.tool === 'StrReplace') {
      const updated = applyStrReplace(files[rel], op.old_string, op.new_string)
      if (updated != null) {
        files[rel] = updated
        lastLine[rel] = op.line
        if (NO_RESUBMIT_PATHS.has(rel) && !hasResubmitContent(updated)) {
          noResubmit[rel] = updated
        }
      }
    }
  }

  const discovered = [...new Set(allOps.map((o) => o.path).filter((p) => p.startsWith(STUDENT_PREFIX)))].sort()
  const targets = new Set([...EXPLICIT_TARGETS, ...discovered])

  const written = []
  const failed = []
  const skippedResubmit = []

  for (const rel of [...targets].sort()) {
    let content
    if (NO_RESUBMIT_PATHS.has(rel) && noResubmit[rel]) {
      content = noResubmit[rel]
      if (files[rel] && hasResubmitContent(files[rel])) skippedResubmit.push(rel)
    } else if (files[rel] != null) {
      content = files[rel]
    } else {
      failed.push(rel)
      continue
    }

    const outPath = path.join(WORKSPACE, ...rel.split('/'))
    fs.mkdirSync(path.dirname(outPath), { recursive: true })
    fs.writeFileSync(outPath, content.replace(/\r?\n/g, '\n'), 'utf8')
    written.push(rel)
  }

  console.log('=== Recovery Report ===')
  console.log(`Transcript lines processed: ${revertLine ?? 'all'}`)
  console.log(`Total tool ops matched: ${allOps.length}`)
  console.log(`Discovered student routes: ${discovered.length ? discovered.join(', ') : '(none)'}`)
  console.log('')
  console.log(`Successfully written (${written.length}):`)
  for (const p of written) {
    const note = skippedResubmit.includes(p) ? ' [used pre-resubmit version]' : ''
    console.log(`  - ${p} (last op ~line ${lastLine[p] ?? '?'})${note}`)
  }
  console.log('')
  if (failed.length) {
    console.log(`Could NOT recover (${failed.length}):`)
    for (const p of failed) console.log(`  - ${p}`)
  } else {
    console.log('Could NOT recover: (none)')
  }
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})
